src/jattach/fdtransfer_server.cpp

```diff
@@ -0,0 +1,334 @@
+            break;
+        }
+    }
+
+    return false;
+}
+
+bool FdTransferServer::sendFd(int fd, struct fd_response *resp, size_t resp_size) {
+    struct msghdr msg = {0};
+
+    struct iovec iov[1];
+    iov[0].iov_base = resp;
+    iov[0].iov_len = resp_size;
+    msg.msg_iov = iov;
+    msg.msg_iovlen = ARRAY_SIZE(iov);
+
+    union {
+       char buf[CMSG_SPACE(sizeof(fd))];
+       struct cmsghdr align;
+    } u;
+
+    if (fd != -1) {
+        msg.msg_control = u.buf;
+        msg.msg_controllen = sizeof(u.buf);
+
+        struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+        cmsg->cmsg_level = SOL_SOCKET;
+        cmsg->cmsg_type = SCM_RIGHTS;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(fd));
+        memcpy(CMSG_DATA(cmsg), &fd, sizeof(fd));
+    }
+
+    ssize_t ret = sendmsg(_peer, &msg, 0);
+    if (ret < 0) {
+        perror("sendmsg()");
+        return false;
+    }
+
+    return true;
+}
+
+static int single_pid_server(pid_t pid) {
+    // get its nspid prior to moving to its PID namespace.
+    pid_t nspid = -1;
+    uid_t _target_uid;
+    gid_t _target_gid;
+    if (!get_process_info(pid, &_target_uid, &_target_gid, &nspid)) {
+        fprintf(stderr, "Process %d not found\n", pid);
+        return 1;
+    }
+
+    if (nspid < 0) {
+        nspid = alt_lookup_nspid(pid);
+    }
+
+    struct sockaddr_un sun;
+    socklen_t addrlen;
+    if (!socketPathForPid(nspid, &sun, &addrlen)) {
+        fprintf(stderr, "Path too long\n");
+        return 1;
+    }
+
+    // Create the server before forking, so w're ready to accept connections once our parent
+    // exits.
+
+    if (!enter_ns(pid, "net")) {
+        fprintf(stderr, "Failed to enter the net NS of target process %d\n", pid);
+        return 1;
+    }
+
+    if (!FdTransferServer::bindServer(&sun, addrlen)) {
+        return 1;
+    }
+
+    if (!enter_ns(pid, "pid")) {
+        fprintf(stderr, "Failed to enter the PID NS of target process %d\n", pid);
+        return 1;
+    }
+
+    // CLONE_NEWPID affects children only - so we fork here.
+    if (0 == fork()) {
+        return FdTransferServer::acceptPeer(&nspid) && FdTransferServer::serveRequests(nspid) ? 0 : 1;
+    } else {
+        // Exit now, let our caller continue.
+        return 0;
+    }
+}
+
+static int path_server(const char *path) {
+    struct sockaddr_un sun;
+    socklen_t addrlen;
+
+    if (!socketPath(path, &sun, &addrlen)) {
+        fprintf(stderr, "Path '%s' is too long\n", path);
+        return 1;
+    }
+
+    if (!FdTransferServer::bindServer(&sun, addrlen)) {
+        return 1;
+    }
+
+    printf("Server ready at '%s'\n", path);
+
+    while (1) {
+        pid_t peer_pid = 0;
+        if (!FdTransferServer::acceptPeer(&peer_pid)) {
+            return 1;
+        }
+
+        // At least a single fork() is need, to actually move a PID namespace.
+        // We fork() twice, so child is automatically repead and we don't have to wait() it.
```

Comment:
Double fork helps to automatically reap the grandchild, but the parent still needs to `waitpid` for the first child.

IMO, `SA_NOCLDWAIT` and just one fork() is a more straghtforward way to handle children.

Reply:
> but the parent still needs to waitpid for the first child.

Correct.

About the forks - there was another reason I preferred the double fork: we fork once, then enter the PID NS which changes our `pid_for_children` namespace, then fork again to have it in effect in the `pid` namespace. Now, our parent still holds a reference to the `pid_for_children` of the target process. As long as its running, that process will keep a reference to the PID NS of the last target process it has accepted, which probably prevents the kernel being able to kill that namespace in case e.g the relevant container goes down. I'm not sure what that might cause (I'm testing with Docker and it doesn't seem like Docker cares about it, but I can definitely see `pid_for_children -> 'pid:[4026532359]` pointing to a "stale" PID NS)

Thinking about it, I can just move back to my original namespace after the fork. Will do that.